### PR TITLE
Add `vue/no-child-content` rule

### LINKED ISSUE #1701: Rule proposal: vue/v-text-no-child-content
**Please describe what the rule should do:**

When using the `v-text` directive, the element's inner content gets overwritten by the result of the `v-text` expression.

**What category should the rule belong to?**

[ ] Enforces code style (layout)
[x] Warns about a potential error (problem)
[ ] Suggests an alternate way of doing something (suggestion)
[ ] Other (please specify:)

This might just belong to the layout category, not sure.

**Provide 2-3 code examples that this rule should warn about:**

```vue
<template>
  <!-- ✓ GOOD -->
  <div v-text="foo"></div>

  <!-- ✗ BAD -->
  <div v-text="foo">{{ bar }}</div>
</template>
```

**Additional context**

I'm in the process of localizing an app, and stumbled upon a few elements which already used `v-text`, but *also* had child content:

```vue
<template>
  <label class="control-label col-sm-3" v-text="fields.name.label">
    Name
  </label>
</template>
```

This isn't really a problem, just unnecessary and confusing.

### PR DESCRIPTION
**Pull request by FloEdelmann**
Fixes #1701.

### CODE REVIEW

**Review by ota-meshi — CHANGES_REQUESTED** [review]
Thank you for this PR!
Would you please add test cases that includes line breaks, indents, and comments in the content? I think these should be allowed.

File: tests/lib/rules/no-child-content.js [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+          suggestions: [
+            { output: '<template><div v-html="foo"></div></template>' }
+          ]
+        }
+      ]
+    },
+    {
+      // v-html directive and whitespace-only text content
+      filename: 'test.vue',
+      code: '<template><div v-html="foo"> </div></template>',
```
Comment (ota-meshi) [inline tests/lib/rules/no-child-content.js:76]: I just noticed this test case 😓.
I think it's better not to report in the case of whitespace or indentation. It's consistent with the vue compiler.


https://sfc.vuejs.org/#eyJBcHAudnVlIjoiPHNjcmlwdCBzZXR1cD5cbmltcG9ydCB7IHJlZiB9IGZyb20gJ3Z1ZSdcblxuY29uc3QgZm9vID0gcmVmKCdIZWxsbyBXb3JsZCEnKVxuPC9zY3JpcHQ+XG5cbjx0ZW1wbGF0ZT48ZGl2IHYtaHRtbD1cImZvb1wiPiA8L2Rpdj48L3RlbXBsYXRlPiIsImltcG9ydC1tYXAuanNvbiI6IntcbiAgXCJpbXBvcnRzXCI6IHtcbiAgICBcInZ1ZVwiOiBcImh0dHBzOi8vc2ZjLnZ1ZWpzLm9yZy92dWUucnVudGltZS5lc20tYnJvd3Nlci5qc1wiXG4gIH1cbn0ifQ==

**Comment by ota-meshi** [comment]
I thought we should allow comments, but the vue compiler doesn't seem to allow comments. It may be better for the rule to report comments as well. What do you think?

https://sfc.vuejs.org/#eyJBcHAudnVlIjoiPHNjcmlwdCBzZXR1cD5cbmltcG9ydCB7IHJlZiB9IGZyb20gJ3Z1ZSdcblxuY29uc3QgZm9vID0gcmVmKCdIZWxsbyBXb3JsZCEnKVxuPC9zY3JpcHQ+XG5cbjx0ZW1wbGF0ZT48ZGl2IHYtaHRtbD1cImZvb1wiPjwhLS1jb21tZW50LS0+PC9kaXY+PC90ZW1wbGF0ZT4iLCJpbXBvcnQtbWFwLmpzb24iOiJ7XG4gIFwiaW1wb3J0c1wiOiB7XG4gICAgXCJ2dWVcIjogXCJodHRwczovL3NmYy52dWVqcy5vcmcvdnVlLnJ1bnRpbWUuZXNtLWJyb3dzZXIuanNcIlxuICB9XG59In0=

File: tests/lib/rules/no-child-content.js (hunk 1 above)
Comment (FloEdelmann) [reply tests/lib/rules/no-child-content.js:76]: That's a helpful website. Let's allow whitespace contents then, and report comments, to stay consistent with the Vue compiler. I'll add the relevant test cases.

File: lib/rules/no-child-content.js [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+      /** @param {VDirective} directiveNode */
+      'VAttribute[directive=true]'(directiveNode) {
+        const directiveName = directiveNode.key.name.name
+        const elementNode = directiveNode.parent.parent
+
+        const elementComments = templateComments.filter((comment) =>
+          inRange(elementNode.range, comment)
+        )
+
+        const childNodes = [...elementNode.children, ...elementComments]
```
Comment (ota-meshi) [inline lib/rules/no-child-content.js:129]: I think you can use the TokenStore to extract all of the content tokens as follows:

```js
const tokenStore = context.parserServices.getTemplateBodyTokenStore()
const childContentTokens = tokenStore.getTokensBetween(elementNode.startTag, elementNode.endTag, { includeComments: true, filter:token => token.type !== 'HTMLWhitespace' })
```

**Review by ota-meshi — CHANGES_REQUESTED** [review]

File: lib/rules/no-child-content.js (hunk 2 above)
Comment (FloEdelmann) [reply lib/rules/no-child-content.js:136]: I simplified the retrieval of comments with this function. However, completely replacing the logic with the tokens returned by this function seemed more effort than it's worth (the suggested fix then doesn't remove whitespace e.g. in `{{ foo }}` anymore). See [`9566ee4`](https://github.com/vuejs/eslint-plugin-vue/pull/1707/commits/9566ee4a20fffdb13feca7cf7c5db58529971504).

Comment (ota-meshi) [reply lib/rules/no-child-content.js:136]: I think your method is better. thank you!

**Review by ota-meshi — APPROVED** [review]
LGTM! Thank you!

**Pull request merged**